python_ga/selection.py

```diff
@@ -0,0 +1,149 @@
+    Pozwala to na lepszy wybor kandydatow, kiedy wartosci funkcji dopasowania osobnikow są duże ale ich wariancje małe
+
+    1. Odwracamy wartosci funkcji dopasowania otrzymując 'inversed_fitness'
+    2. Obliczamy 'fitness_dispersion' rozstęp wartość funkcji dopasowania osobników w populacji
+    3. Obliczamy 'fitness_dispersion_ratio' stosunek wartości rozstępu do średniej wartości funkcji dopasowania
+    4. Wyznaczamy znormalizowane wartosci funkcji dopasowania osobnikow 'dispersed_fitness' poprzez odjęcie od nich
+    wartości najgorszego osobnika, pomnozonego przez 1 - 'fitness_dispersion_ratio'
+    5. Obliczamy listę 'wheel_probability_list' będącą sumą skumulowaną 'dispersed_fitness' wszystkich osobników
+    6. Losujemy 'select_n' osobników używając listy 'wheel_probability_list'.
+
+    :param population:  list
+    :param pop_fitness: list
+    :param select_n:    int
+    :param config:      dict
+    :return: list
+    """
+    selected = []
+
+    if 'candidates_dispersion' in config and config['candidates_dispersion']:
+        fitness_dispersion = max(pop_fitness) - min(pop_fitness)
+        fitness_dispersion_ratio = fitness_dispersion / np.mean(pop_fitness)
+        fitness = pop_fitness - min(pop_fitness) * (1 - fitness_dispersion_ratio)
+    else:
+        fitness = pop_fitness
+    wheel_probability_list = np.cumsum(fitness) / sum(fitness)
+
+    for i in range(select_n):
+        r = np.random.random()
+        for j, proba in enumerate(wheel_probability_list):
+            if r < proba:
+                selected.append(population[j])
+                break
+    return selected
+
+
+def simple_selection(population, pop_fitness, select_n, config):
+    """
+    Metoda selekcji prostej:
+    1. Indeksy osobników w populacji są sortowane rosnąco względem ich wyników w pop_fitness;
+    2. select_n osobników z najlepszymi wynikami wybieranych jest do puli selekcji;
+    :param population:  list
+    :param pop_fitness: list
+    :param select_n:    int
+    :param config:      dict
+    :return: list
+    """
+    best_index = np.argsort(pop_fitness)[::-1]
+    return [population[i] for i in best_index][:select_n]
+
+
+def tournament_selection_classic(population, pop_fitness, select_n, config):
+    """
+    Metoda selekcji tournament - klasyczna
+    1. Wybieramy len(population / select_n) osobników
+    2. Porównujemy dopasowanie len(population / select_n) osobników
+    3. Wybieramy pojedynczego osobnika, gdzie prawdopobienstwa wyboru wynoszą p dla najlepszego osobnika, p(1 - p)
+     dla drugiego, p(1 - p)^2 dla trzeciego... wykorzystując roulette_selection bez rozpraszania osobników
+    4. Powtarzamy aż uzyskania select_n wybranych osobników.
+    :param population:  list
+    :param pop_fitness: list
+    :param select_n:    int
+    :param config:      dict
+    :return: list
+    """
+    def sort_candidates():
```

Comment:
`sort_candidates` returns sorted candidates and indices of sorted candidates, for the purpose of better and easier debugging

algorithm logic is not changed